### PR TITLE
Bug 1116566 - Create test to catch regression in bug 1116087

### PR DESCRIPTION
**Pull request by mwargers**

### CODE REVIEW

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/manifest.ini [hunk 1]
```diff
@@ -4,6 +4,9 @@
 online = true
 smoketest = true
 
+[test_bug_1116566.py]
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/manifest.ini]: We need a more explicit name here, if this test fails again, it would take longer to know what's the issue.

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+        GaiaTestCase.setUp(self)
+        self.connect_to_local_area_network()
+
+        # Open email app
+        self.email = Email(self.marionette)
+        self.email.launch()
+
+    def test_bug_1116566(self):
+        """ https://bugzilla.mozilla.org/show_bug.cgi?id=1116566 """
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py]: I would leave the link to the actual regression https://bugzilla.mozilla.org/show_bug.cgi?id=1116087 instead.

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        # setup email account
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py]: I don't think this comment adds much value to the line below.

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        self.email.setup_IMAP_email(self.testvars['email']['IMAP'])
+
+        # check account has emails
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py]: I don't think this comment adds much value to the line below.

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        self.email.wait_for_emails_to_sync()
+        self.assertGreater(len(self.email.mails), 0)
+
+        self.assertEqual(len(self.marionette.find_elements(*self._menu_button_locator)), 1, 'Should have only 1 menu-button')
+
+        # Pull up the cards view
+        self.device.hold_home_button()
+        cards_view = CardsView(self.marionette)
+        cards_view.wait_for_cards_view()
+        cards_view.wait_for_card_ready('email')
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py]: I think using `self.email.name` would be more appropriate. It'd help if we change the name of the app one day (this happened with the homescreen last summer)

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        cards_view.close_app('email')
+
+        self.marionette.switch_to_frame()
+        self.email.launch()
+        # check account has emails
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py]: I don't think this comment adds much value to the line below.

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    def setUp(self):
+        try:
+            self.testvars['email']['IMAP']
+        except KeyError:
+            raise SkipTest('account details not present in test variables')
+
+        GaiaTestCase.setUp(self)
+        self.connect_to_local_area_network()
+
+        # Open email app
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py]: I don't think this comment adds much value to the line below.

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        self.email = Email(self.marionette)
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py]: What's the advantage of launching the app in `setUp` here?

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+from marionette import SkipTest
+from marionette.by import By
+
+from gaiatest import GaiaTestCase
+from gaiatest.apps.email.app import Email
+from gaiatest.apps.system.regions.cards_view import CardsView
+
+
+class TestBug1116566(GaiaTestCase):
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py]: Same comment regarding the file name

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    _menu_button_locator = (By.CSS_SELECTOR, '.msg-folder-list-btn')
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py]: I agree that this test case is unusual. I think we can do better here. 

I guess we should use the header locator instead as the problem concerned the whole header. This made me notice that the [header class](https://github.com/mozilla-b2g/gaia/blob/master/tests/python/gaia-ui-tests/gaiatest/apps/email/app.py#L199) has access to the entire page. I think we should restrict its scope like we do [here for instance](https://github.com/mozilla-b2g/gaia/blob/master/tests/python/gaia-ui-tests/gaiatest/apps/contacts/app.py#L124).

In other words, if we restrict the header to a [PageRegion](https://github.com/mozilla-b2g/gaia/blob/master/tests/python/gaia-ui-tests/gaiatest/apps/base.py#L133), we would have to define the locator of the header (`By.CSS_SELECTOR, .msg-list-header`) inside the Email page class. Then we would be able to use it in the test this unsual call `len(self.marionette.find_elements(email.header_locator))`.

What do you think about that?

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py (hunk 2 above)
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py]: Yeah, I agree. I don't even think that a special bug had to be created anyway, to make a gaia-ui test out of this. The gaia-ui test writing could just be done in the bug itself, as is normally done with automated test writing, in general.

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py (hunk 3 above)
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py]: Yeah, I copied this from test_IMAP_email_notification.py

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py (hunk 5 above)
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py]: I tried this, this didn't work. The app name is 'E-Mail' in this case, but this api needs the app origin, which is 'app://email.gaiamobile.org'.

The cards_view code doesn't really work well in this case. It takes the app name, then makes it lower case, but in this case, it also would have to get rid of the '-' sign:
http://mxr.mozilla.org/gaia/source/tests/python/gaia-ui-tests/gaiatest/apps/system/regions/cards_view.py#24

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py (hunk 8 above)
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py]: Nothing really, this came from test_IMAP_email_notification.py

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py (hunk 10 above)
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py]: Ok, sounds good. I don't really know what PageRegion is doing, currently, to be hones.

Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_bug_1116566.py]: Ah, I know it know.

**Comment by try-server-hook** [comment]
[mwargers Martijn (mwargers)](https://github.com/mwargers) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=d3acd6083226)

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_only_one_header_displayed.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+from marionette import SkipTest
+from marionette.by import By
+
+from gaiatest import GaiaTestCase
+from gaiatest.apps.email.app import Email
+from gaiatest.apps.system.regions.cards_view import CardsView
+
+
+class TestOnlyOneHeaderDisplayed(GaiaTestCase):
+    _menu_button_locator = (By.CSS_SELECTOR, '.msg-folder-list-btn')
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_only_one_header_displayed.py]: Nit: We don't need it anymore

File: tests/python/gaia-ui-tests/gaiatest/apps/email/app.py [hunk 12]
```diff
@@ -196,32 +197,32 @@ def __call__(self, marionette):
                 return False
 
 
-class Header(Base):
+class Header(PageRegion):
 
     _menu_button_locator = (By.CSS_SELECTOR, '.card.center .msg-folder-list-btn')
```
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/apps/email/app.py]: I think we can remove ".card.center" from these 3 locators.

Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/email/app.py]: Yeah, I agree.

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_only_one_header_displayed.py (hunk 11 above)
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_only_one_header_displayed.py]: Oops, forgot that one.

**Comment by try-server-hook** [comment]
[mwargers Martijn (mwargers)](https://github.com/mwargers) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=997069ef54f6)

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_only_one_header_displayed.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        GaiaTestCase.setUp(self)
+        self.connect_to_local_area_network()
+
+        self.email = Email(self.marionette)
+        self.email.launch()
+
+    def test_only_one_header_displayed(self):
+        """ https://bugzilla.mozilla.org/show_bug.cgi?id=1116087 """
+
+
```
Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_only_one_header_displayed.py]: remove one line

File: tests/python/gaia-ui-tests/gaiatest/apps/email/app.py [hunk 14]
```diff
@@ -31,6 +31,7 @@ class Email(Base):
     _search_textbox_locator = (By.CSS_SELECTOR, 'form[role="search"]')
     _email_subject_locator = (By.XPATH, '//a[@data-index!="-1"]/div/span[text()="%s"]')
     _back_button_locator = (By.CLASS_NAME, 'sup-back-btn')
+    header_locator = (By.CSS_SELECTOR, '.msg-list-header')
```
Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/apps/email/app.py]: this should have _ prefix, as it is a locator

File: tests/python/gaia-ui-tests/gaiatest/apps/email/app.py [hunk 15]
```diff
@@ -142,7 +143,7 @@ def a11y_navigate_to_manual_setup(self, name, email):
 
     @property
     def header(self):
-        return Header(self.marionette)
+        return Header(self.marionette, self.marionette.find_element(*self.header_locator))
```
Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/apps/email/app.py]: I would undo this change

File: tests/python/gaia-ui-tests/gaiatest/apps/email/app.py [hunk 16]
```diff
@@ -196,32 +197,32 @@ def __call__(self, marionette):
...
 
-class Header(Base):
+class Header(PageRegion):
 
-    _menu_button_locator = (By.CSS_SELECTOR, '.card.center .msg-folder-list-btn')
-    _compose_button_locator = (By.CSS_SELECTOR, '.card.center .msg-compose-btn')
-    _label_locator = (By.CSS_SELECTOR, '.card.center .msg-list-header-folder-label.header-label')
+    _menu_button_locator = (By.CSS_SELECTOR, '.msg-folder-list-btn')
+    _compose_button_locator = (By.CSS_SELECTOR, '.msg-compose-btn')
+    _label_locator = (By.CSS_SELECTOR, '.msg-list-header-folder-label.header-label')
```
Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/apps/email/app.py]: :+1:  for shorting locators

File: tests/python/gaia-ui-tests/gaiatest/apps/email/app.py (hunk 14 above)
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/apps/email/app.py]: We use this attribute outside of the class. I think we should keep it as public.

File: tests/python/gaia-ui-tests/gaiatest/apps/email/app.py (hunk 15 above)
Comment (JohanLorenzo) [inline tests/python/gaia-ui-tests/gaiatest/apps/email/app.py]: Why do you think it's worth it to remove this change?

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_only_one_header_displayed.py [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+        self.email.setup_IMAP_email(self.testvars['email']['IMAP'])
+        self.email.wait_for_emails_to_sync()
+        self.assertGreater(len(self.email.mails), 0)
+
+        self.assertEqual(len(self.marionette.find_elements(*self.email.header_locator)), 1, 'Should have only 1 list-header')
```
Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_only_one_header_displayed.py]: I believe in this specific case we should define the locator in the test and not in the email app.
That means define the locator above the setUp section : 
<code> _messages_list_header_locator = (By.CSS_SELECTOR, '.msg-list-header') </code>
 and here use <code>self.marionette.find_elements(*self._messages_list_header_locator)</code>

 <code>_messages_list_header_locator</code> is a more suggestive name for this locator

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/manifest.ini [hunk 18]
```diff
@@ -4,6 +4,9 @@
 online = true
 smoketest = true
 
+[test_only_one_header_displayed.py]
```
Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/manifest.ini]: The test name is not suggestive enough, but I can't come with a better solution. I will thinks about it and add a comment.

Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/manifest.ini]: <code>test_email_one_header_displayed.py</code>

File: tests/python/gaia-ui-tests/gaiatest/apps/email/app.py (hunk 14 above)
Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/apps/email/app.py]: Ok, got it!

Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/apps/email/app.py]: can we just rename this  to: <code>emails_list_header_locator</code>?

File: tests/python/gaia-ui-tests/gaiatest/apps/email/app.py (hunk 15 above)
Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/apps/email/app.py]: Actually, leave it as it is. We need this, as PageRegion takes the root element as parameter.

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/manifest.ini (hunk 18 above)
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/manifest.ini]: Fine by me. This test is already in the email/ directory, though, so I would think it's already pretty clear that this is an email test.

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_only_one_header_displayed.py (hunk 17 above)
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_only_one_header_displayed.py]: Ok, will do.

Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_only_one_header_displayed.py]: Note that there should be only 1 header displayed, so email_headers_displayed is not a good name. I'll use email_header.

**Comment by try-server-hook** [comment]
[mwargers Martijn (mwargers)](https://github.com/mwargers) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=a15c6353ee57)

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_email_one_header_displayed.py [hunk 19]
```diff
@@ -0,0 +1,50 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from marionette import SkipTest
+from marionette.by import By
```
Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_email_one_header_displayed.py]: unused import

**Comment by viorelaioia** [comment]
PR looks good.
Though, I ran the test locally on b2g desktop, and it failed constantly with 'JavascriptException: JavascriptException: TypeError: appWindow is undefined', when trying to re-launch Email app.
I'm not sure why are we getting that error. Locally, I added a wait for homescreen to be displayed after Email app was closed, and then re-launch it. That wait seems to fix this issue, as the test didn't failed again 
<code>self.wait_for_condition(lambda m: self.apps.displayed_app.name == Homescreen.name)</code>

**Comment by mwargers** [comment]
Thanks! I'll give it a try and use that if it works.

**Comment by try-server-hook** [comment]
[mwargers Martijn (mwargers)](https://github.com/mwargers) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=33de44d03702)

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_email_one_header_displayed.py [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+        self.assertGreater(len(self.email.mails), 0)
+
+        email_header_list = self.marionette.find_elements(*self.email.emails_list_header_locator)
+        self.assertEqual(len(email_header_list), 1, 'Should have only 1 list-header')
+
+        self.device.hold_home_button()
+        cards_view = CardsView(self.marionette)
+        cards_view.wait_for_cards_view()
+        cards_view.wait_for_card_ready(self.email.name)
+        cards_view.close_app(self.email.name)
```
Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_email_one_header_displayed.py]: Here you can add an assert that the app is no longer presents, as in [test_cards_view_kill_apps_with_two_apps.py][link]
[link]:https://github.com/mozilla-b2g/gaia/blob/master/tests/python/gaia-ui-tests/gaiatest/tests/functional/cards_view/test_cards_view_kill_apps_with_two_apps.py#L40

File: tests/python/gaia-ui-tests/gaiatest/apps/system/regions/cards_view.py [hunk 21]
```diff
@@ -72,6 +73,9 @@ def close_app(self, app):
         self.wait_for_card_ready(app)
         self.marionette.find_element(*self._close_button_locator(app)).tap()
         self.wait_for_element_not_present(*self._app_card_locator(app))
+        if len(self.cards) == 0:
+            self.marionette.switch_to_frame()
+            self.wait_for_condition(lambda m: self.apps.displayed_app.name == Homescreen.name)
```
Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/apps/system/regions/cards_view.py]: I think this condition should be added in the test, along with the import, as in [test_sms_auto_save_draft.py][link]
[link]:https://github.com/mozilla-b2g/gaia/blob/master/tests/python/gaia-ui-tests/gaiatest/tests/functional/messages/test_sms_auto_save_draft.py#L46

**Comment by viorelaioia** [comment]
Comments asside, lgtm.

File: tests/python/gaia-ui-tests/gaiatest/apps/system/regions/cards_view.py (hunk 21 above)
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/system/regions/cards_view.py]: I'll make the change, but I'm not sure I agree with this.
I'd like to see these kinds of methods (in the app/pageregion objects) as something that's doing a UI action or a programmatic action and then it would make sure that it's in a stable state afterwards.
In case there are no cards left with the close_app method, the device returns to the Homescreen app. In my opinion, this method should wait for that state to happen.

Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/apps/system/regions/cards_view.py]: Otoh, there are lots of other methods that are not really doing that, for instance self.device.hold_home_button() where you have to use a bunch of stuff to make sure the cards_view is ready.

Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/apps/system/regions/cards_view.py]: I still believe we should add this condition in the test. To me, it is part of the test logic and it should be in the test itself, rather than in the app object. We've been using this condition in several other tests, like [test_ftu_skip_tour.py
][link] or [test_ftu_with_tour.py][link]
[link]:https://github.com/mozilla-b2g/gaia/blob/master/tests/python/gaia-ui-tests/gaiatest/tests/functional/ftu/test_ftu_skip_tour.py#L111
[link]:https://github.com/mozilla-b2g/gaia/blob/master/tests/python/gaia-ui-tests/gaiatest/tests/functional/ftu/test_ftu_with_tour.py#L47

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_email_one_header_displayed.py (hunk 20 above)
Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_email_one_header_displayed.py]: I''l be using this:
        self.assertEqual(len(cards_view.cards), 0, 'Should have no cards left to display')
Since there are no cards left to display anymore.

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_email_one_header_displayed.py [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+
+        self.assertEqual(len(cards_view.cards), 0, 'Should have no cards left to display')
+        self.marionette.switch_to_frame()
```
Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_email_one_header_displayed.py]: no need for this switch_to_frame()

File: tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_email_one_header_displayed.py (hunk 20 above)
Comment (viorelaioia) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_email_one_header_displayed.py]: I believe we'd better check that Email app no longer appears in the cards view(it is closed), rather than check that there are no cards view displayed, though it's the same thing

Comment (mwargers) [inline tests/python/gaia-ui-tests/gaiatest/tests/functional/email/test_email_one_header_displayed.py]: Ok, I'm adding a check for that too.
